Objects/dictobject.c

```diff
@@ -6819,38 +6993,98 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
     return 0;
 }
 
+static bool
+set_dict_inline_values(PyObject *obj, PyObject *new_dict)
+{
+    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(obj);
+
+    PyDictValues *values = _PyObject_InlineValues(obj);
+
+#ifdef Py_GIL_DISABLED
+    PyDictObject *dict = _PyObject_ManagedDictPointer(obj)->dict;
+    if (dict != NULL) {
+        // We lost a race with materialization of the dict
+        return false;
+    }
+#endif
+
+    Py_XINCREF(new_dict);
+    _PyObject_ManagedDictPointer(obj)->dict = (PyDictObject *)new_dict;
+
+    if (values->valid) {
+        FT_ATOMIC_STORE_UINT8_RELAXED(values->valid, 0);
+        for (Py_ssize_t i = 0; i < values->capacity; i++) {
+            Py_CLEAR(values->values[i]);
+        }
+    }
+
+    return true;
+}
+
 void
-PyObject_ClearManagedDict(PyObject *obj)
+_PyObject_SetManagedDict(PyObject *obj, PyObject *new_dict)
```

Comment:
This function (and related helper functions) could use some clean-up:

- Let's avoid the weird `goto` if possible
- The stores to `_PyObject_ManagedDictPointer(obj)->dict` need to be atomic
- I think it'll be clearer if you lift the fallible check in `set_dict_inline_values` into this function
- The assertions in `_PyDict_DetachFromObject` may not hold: once you lock the dict and object, the dict may no longer be attached.
- We need to assign `new_dict` before we mark the values as invalid